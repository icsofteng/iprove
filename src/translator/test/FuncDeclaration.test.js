const {translate} = require('../../translator/z3')
const {translate: translate_latex} = require('../../translator/latex')

const test_constants = [{ value: 'P', varType: 'Bool' }]
const test_rules = [
  {
    type: 'funcDef',
    name: 'friends',
    params: [{type: 'type', value: 'Person'}, {type: 'type', value: 'Person'}],
    returnType: {type:'type', value: 'Bool'}
  },
  {
    type: 'literal',
    value: 'P'
  }
]

test('func Def test', () => {
  expect(translate(test_rules, test_constants, [], ['Person'], [])).toMatchSnapshot()
})

const test_constants2 = [{ value: 'P', varType: 'Bool' }]
const test_rules2 = [
  {
    type:'funcDef',
    name:'Power',
    params:[{type: 'type', value: 'Dragon'}, {type: 'type', value: 'Dragon'}],
    returnType: {type:'type', value: 'Int'}
  },
  {
    type: 'literal',
    value: 'P'
  }
]
test('func Def Dragon test', () => {
  expect(translate(test_rules2, test_constants2, [], ['Dragon'], [])).toMatchSnapshot()
})

const test_constants3 = [{ value: 'P', varType: 'Bool' }]
const test_rules3 = [
  {
    type:'funcDef',
    name:'Happy',
    params:[{type: 'type', value: 'Dragon'}, {type: 'type', value: 'Human'}],
    returnType: {type:'type', value: 'Reward'}
  },
  {
    type: 'literal',
    value: 'P'
  }
]
test('func Def Happy test', () => {
  expect(translate(test_rules3, test_constants3, [], ['Reward' ,'Dragon', 'Human'], [])).toMatchSnapshot()
})


const func_def_test_rules = [
  {
    type: 'funcDef',
    name: 'friends',
    params: [
        {
          type: 'type',
          value: 'Person'
        },
        {
          type: 'type',
          value: 'Person'
        }
    ],
    returnType: {
        type: 'type',
        value: 'Bool'
    }
  }
]
test('FuncDef test latex', () => {
  expect(translate_latex(func_def_test_rules)[0]).toMatchSnapshot()
})

