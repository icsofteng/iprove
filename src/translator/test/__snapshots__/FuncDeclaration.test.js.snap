// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FuncDef test latex 1`] = `"define \\\\ friends(Person, Person): Bool"`;

exports[`func Def Dragon test 1`] = `
"(declare-sort Dragon)
(declare-const P Bool)
(assert (not P))
(check-sat)"
`;

exports[`func Def Happy test 1`] = `
"(declare-sort Reward)
(declare-sort Dragon)
(declare-sort Human)
(declare-const P Bool)
(assert (not P))
(check-sat)"
`;

exports[`func Def test 1`] = `
"(declare-sort Person)
(declare-const P Bool)
(assert (not P))
(check-sat)"
`;
